docs/query-metadata-style-guide.md

```diff
@@ -0,0 +1,173 @@
+
+## Query files (.ql extension)
+
+Query files have the extension `.ql`. Each file has two distinct areas:
+
+*   Metadata area–displayed at the top of the file, contains the metadata that defines how results for the query are interpreted and gives a brief description of the purpose of the query.
+*   Query definition–defined using QL. The query includes a select statement, which defines the content and format of the results. For further information about writing QL, see the following topics:
+    *   [Learning QL](https://help.semmle.com/wiki/display/QL/Learning+QL)
+    *   [QL language handbook](https://help.semmle.com/QL/ql-handbook/index.html)
+    *   [QL language specification](https://help.semmle.com/QL/QLLanguageSpecification.html)
+    *   [QL style Guide](https://github.com/Semmle/ql/blob/master/docs/ql-style-guide.md) 
+
+For examples of query files for the languages supported by Semmle, see the following pages: 
+
+*   [C/C++ queries](https://wiki.semmle.com/pages/viewpage.action?pageId=19334052)
+*   [C# queries](https://wiki.semmle.com/display/CSHARP/C%23+queries)
+*   [COBOL queries](https://wiki.semmle.com/display/COBOL/COBOL+queries)
+*   [Java queries](https://wiki.semmle.com/display/JAVA/Java+queries)
+*   [JavaScript queries](https://wiki.semmle.com/display/JS/JavaScript+queries)
+*   [Python queries](https://wiki.semmle.com/display/PYTHON/Python+queries)
+
+## Metadata area
+
+Query file metadata contains important information which defines the name and purpose of the query. The metadata is included as the content of a valid QLDoc comment, on lines with leading whitespace followed by `*`, between an initial `/**` and a trailing `*/`. For example:
+
+```
+/**
+ * @name Useless assignment to local variable
+ * @description An assignment to a local variable that is not used later on, or whose value is always
+ *              overwritten, has no effect.
+ * @kind problem
+ * @problem.severity warning
+ * @id cs/useless-assignment-to-local
+ * @tags maintainability
+ *       external/cwe/cwe-563
+ * @precision very-high
+ */
+ ```
+
+In order to help others use your query, and to ensure that the query works correctly on LGTM.com, you should include all of the required information outlined below in the metadata, and as much of the optional information as possible. For further information on query metadata see [Query file requirements](https://help.semmle.com/wiki/display/SD/Query+file+requirements).
+
+### Query name `@name`
+
+You must specify an `@name` property for your query. This property defines the display name for the query. Query names should use sentence capitalization, but not include a full stop. Filter queries should specify the results that are excluded when you run the query. See the following examples for more detail:
+
+*   `@name Access to variable in enclosing class`
+*   `@name Array argument size mismatch`
+*   `@name Reference equality test on strings`
+*   `@name Return statement outside function`
```

Comment:
Should we include an example of a filter and a metric name here, or is it fine to concentrate on problem queries?

Reply:
I suppose for completeness they should be included, but discussed metric queries in particular with various developers and we decided to concentrate on problems for now - it seems a bit strange to give advice on how to write metadata for queries whose results are not shown for most users.

Reply:
This comment was triggered by the inclusion of filter in the paragraph above. If we're going to focus on alert queries, then it may be best to omit "filter" from this section for now.

Reply:
Good point!